### PR TITLE
feat(ResourceTable): randomize rid generation

### PR DESCRIPTION
**Pull request by ecyrbe**
This PR adds randomization of resource id generation.
The downside: it adds rand crate to core.

### CODE REVIEW

**Comment by bartlomieju** [comment]
@ecyrbe just a heads up: in `cli/js/files.ts` we create `stdin/out/err` and they rely on resource id being incrementing integers - it needs to be updated otherwise `js_unit_test` will fail.

**Comment by ecyrbe** [comment]
> stdin/out/er

oh, i see. thanks. i'll fix it.

**Comment by ecyrbe** [comment]
@bartlomieju i made it so we maintain compatibility with this behaviour. if you want i can go the other way around and add an op to get the default stdin/stdout/stderr from typescript side. your call.

File: core/resources.rs [hunk 1]
```diff
@@ -42,11 +43,14 @@ impl ResourceTable {
...
   }
 
-  // TODO: resource id allocation should probably be randomized for security.
+  // resource id allocation are randomized for security.
   fn next_rid(&mut self) -> ResourceId {
-    let next_rid = self.next_id;
-    self.next_id += 1;
-    next_rid as ResourceId
+    let mut rng = rand::thread_rng();
+    let mut next_rid = rng.gen::<u32>() & 0x7FFF_FFFF;
```
Comment (ry) [inline core/resources.rs]: I guess ` 0x7FFF_FFFF` is the largest integer allowed in JS? I think this deserves a comment.

File: core/resources.rs [hunk 2]
```diff
@@ -56,6 +60,16 @@ impl ResourceTable {
     rid
   }
 
+  pub fn placement_add(
```
Comment (ry) [inline core/resources.rs]: Please add a triple slash comment explaining purpose.

File: core/resources.rs [hunk 3]
```diff
@@ -79,3 +93,67 @@ impl ResourceTable {
...
+  fn test_remove_from_resource_table() {
+    let mut table = ResourceTable::default();
+    let rid1 = table.add("fake1", Box::new(FakeResource { not_empty: 1 }));
+    let rid2 = table.add("fake2", Box::new(FakeResource { not_empty: 2 }));
+    assert_eq!(table.map.len(), 2);
+    table.close(rid1);
+    assert_eq!(table.map.len(), 1);
+    table.close(rid2);
+    assert_eq!(table.map.len(), 0);
+  }
```
Comment (ry) [inline core/resources.rs]: Nice tests

File: core/Cargo.toml [hunk 4]
```diff
@@ -19,6 +19,7 @@ futures = { version = "0.3.1", features = [ "thread-pool", "compat" ] }
 lazy_static = "1.4.0"
 libc = "0.2.66"
 log = "0.4.8"
+rand = "0.7.2"
```
Comment (ry) [inline core/Cargo.toml]: Unfortunate, but I'm ok with it.

File: core/resources.rs (hunk 1 above)
Comment (ecyrbe) [reply core/resources.rs]: In fact, no. But rid's are cast into i32 then interpreted as unsigned number on typescript side.
Then when sending rid's back to Rust the cast back to i32 fails as it overflows.
There are two options :
- the one i took and make the generated rid a valid positive i32
- make rid u32 on Ops api side. But there might be a good reason someone made them i32, and i did not have the Time to investigate why

Comment (ecyrbe) [reply core/resources.rs]: i actually need feedback on this, so i'll not resolve

**Comment by kevinkassimo** [comment]
Actually out of curiosity, is there a reason we want to make rid generation random?

I feel that from 0 to 0x7FFFFFFF is actually not that hard to fully iterate, so from security point of view this does not bring too much of a benefit if we would ever want to use the rid as a mechanism to avoid unwanted access to resources by malicious code.

**Comment by ry** [comment]
@kevinkassimo Less for security, more because we shouldn't be writing code that relies on specific rid numbers. All else being the same, we might as well make it random.

**Comment by ecyrbe** [comment]
@kevinkassimo also never underestimate security issues ans how they will be exploited. Iterating over 2 billion rid from a network API exposed by deno web service ( round trip time, complexity of the exploit to gain access to deno rid from said API etc ) will prove to already be a hard job, but making rid random will make such exploit unfeasable. but i'm all for making them u64 and random. More security.

**Comment by kevinkassimo** [comment]
@ecyrbe Well, it is not that hard for a long running process such as a server to give code enough time to iterate. Once malicious code gets its way to the server, it can be patient, and rid space smaller than 2^64 provides no benefit IMO (cryptographically often 2^80 is the starting point where stuff can be viewed as secure).

I do agree that making code less dependent on specific rid numbers is a good idea. Generating random number should not be too huge a performance cost so I am okay with it (I guess we don't need to worry wasting too much time on collisions, since maintaining a huge set of rids is unlikely)

**Comment by ecyrbe** [comment]
@kevinkassimo my proposal already takes Care about collisions.

**Comment by kevinkassimo** [comment]
@ecyrbe I knew that. I mean the case where colliding multiple times before finding a free rid (e.g. the unrealistic scenario where only 1 rid is free and all others occupied, under which basically spending all the time colliding) (which is unlikely as I said before)

**Comment by ecyrbe** [comment]
@kevinkassimo I do agree. Thanks for your feedback, it's really appreciated. I guess i should have opened an issue to have these kind of discussion before and not after impl. if we don't want it, we should at least take the tests part of it as there was no tests for ResourceTable before this PR.

File: core/resources.rs [hunk 5]
```diff
@@ -42,11 +43,17 @@ impl ResourceTable {
...
   fn next_rid(&mut self) -> ResourceId {
-    let next_rid = self.next_id;
-    self.next_id += 1;
-    next_rid as ResourceId
+    let mut rng = rand::thread_rng();
+    // rid's are cast into i32 then interpreted as unsigned number on typescript side.
+    // then when sending rid's back to Rust the cast back to i32 fails as it overflows.
+    // here we make the generated rid a valid positive i32 to prevent overflow
+    let mut next_rid = rng.gen::<u32>() & 0x7FFF_FFFF;
+    while self.map.contains_key(&next_rid) {
```
Comment (ry) [inline core/resources.rs]: I think it should panic if there is a collision.

File: core/resources.rs [hunk 6]
```diff
@@ -42,11 +43,17 @@ impl ResourceTable {
     None
   }
 
-  // TODO: resource id allocation should probably be randomized for security.
+  // resource id allocation are randomized for security.
   fn next_rid(&mut self) -> ResourceId {
-    let next_rid = self.next_id;
-    self.next_id += 1;
-    next_rid as ResourceId
+    let mut rng = rand::thread_rng();
```
Comment (ry) [inline core/resources.rs]: Does this obey `--seed`?

Comment (ecyrbe) [reply core/resources.rs]: no, it uses system generated seed.

File: core/resources.rs (hunk 5 above)
Comment (ecyrbe) [reply core/resources.rs]: why? this would make the whole system unstable.

**Comment by ecyrbe** [comment]
no activity since last update.
closing.

**Comment by aligoren** [comment]
Hi there. Is this still in the todo?

```rust
use rand::RngCore;

fn next_rid() -> u32 {

    let mut rng = rand::rngs::OsRng;

    let next_rid = rng.next_u32();


    next_rid as u32
}
```

Why this is unacceptable?

Thanks!

**Pull request closed without merging**